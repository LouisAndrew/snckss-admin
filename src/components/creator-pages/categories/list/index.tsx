import React from 'react'
import { ListGroup, Label, ListGroupItem } from 'reactstrap'
import { InlineIcon } from '@iconify/react'
import deleteIcon from '@iconify/icons-mdi/delete'

export interface ListItem {
        text: string
        key: any
}

interface Props {
        headerText: string
        items: ListItem[]
        handleRemove: (key: ListItem['key']) => void
}

const List: React.FC<Props> = ({ headerText, items, handleRemove }) => {
        return (
                <ListGroup>
                        <Label>{headerText}</Label>
                        {items.map((item) => (
                                <ListGroupItem key={item.key}>
                                        {item.text}
                                        <span
                                                className="delete"
                                                onClick={() =>
                                                        handleRemove(item.key)
                                                }
                                        >
                                                <InlineIcon icon={deleteIcon} />
                                        </span>
                                </ListGroupItem>
                        ))}
                </ListGroup>
        )
}

export default List
