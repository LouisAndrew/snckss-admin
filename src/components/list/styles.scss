.list-label {
        font-weight: bold;
}

.list-group-item {
        display: flex !important;
        justify-content: space-between;

        .delete {
                transition: 0.2s;

                &:hover {
                        cursor: pointer;
                        transform: scale(1.5);
                }
        }

        .clickable:hover {
                cursor: pointer;
        }
}
