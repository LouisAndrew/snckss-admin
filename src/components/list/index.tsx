import React from 'react'
import { ListGroup, Label, ListGroupItem } from 'reactstrap'
import { InlineIcon } from '@iconify/react'
import deleteIcon from '@iconify/icons-mdi/delete'

import './styles.scss'

export interface ListItem {
        text: string
        key: any
}

interface Props {
        headerText: string
        items: ListItem[]
        handleRemove: (key: ListItem['key']) => void
        handleClick?: (key: ListItem['key']) => void
}

const List: React.FC<Props> = ({
        headerText,
        items,
        handleRemove,
        handleClick,
}) => {
        return (
                <ListGroup>
                        <Label className="list-label">{headerText}</Label>
                        {items.map((item) => (
                                <ListGroupItem
                                        key={item.key}
                                        data-testid="list-items"
                                >
                                        <div
                                                className={`text ${
                                                        handleClick
                                                                ? 'clickable'
                                                                : ''
                                                }`}
                                                onClick={() => {
                                                        if (handleClick) {
                                                                handleClick(
                                                                        item.key
                                                                )
                                                        }
                                                }}
                                        >
                                                {item.text}
                                        </div>
                                        <span
                                                className={`del-${item.text} delete`}
                                                onClick={() =>
                                                        handleRemove(item.key)
                                                }
                                        >
                                                <InlineIcon icon={deleteIcon} />
                                        </span>
                                </ListGroupItem>
                        ))}
                </ListGroup>
        )
}

export default List
