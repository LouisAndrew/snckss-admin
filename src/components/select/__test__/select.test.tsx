import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import { render, cleanup, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

import Select from '..'
import { ListItem } from '../../list'

describe('Select reusable component', () => {
        interface mockItems {
                name: string
        }

        let calledAvailable: mockItems[] = []
        let calledSelected: mockItems[] = []
        const mockAvailable: mockItems[] = [
                {
                        name: '1',
                },
                {
                        name: '2',
                },
                {
                        name: '3',
                },
        ]
        const mockSelected: mockItems[] = [
                {
                        name: '4',
                },
                {
                        name: '5',
                },
        ]
        const mockHeaderText: string = 'Select One'

        const handleChange = jest.fn(
                (event: React.ChangeEvent<HTMLInputElement>) => {
                        const which: string = event.target.value
                        const item: mockItems = mockAvailable.filter(
                                (mck) => mck.name === which
                        )[0]

                        calledAvailable = [...calledAvailable, item]
                }
        )

        // handle remove doesn't need to be tested here bcs handle remove would be
        // passed as the remove function to list component and is already tested
        // separately
        const handleRemove = jest.fn(() => {})

        afterEach(cleanup)
        const el: React.ReactElement = (
                <Select
                        available={mockAvailable}
                        selected={mockSelected}
                        headerText={mockHeaderText}
                        handleChange={handleChange}
                        handleRemove={handleRemove}
                />
        )

        it('renders without crashing', () => {
                const div = document.createElement('div')
                ReactDOM.render(el, div)
        })

        it('renders the available options correctly', () => {
                const { getByText } = render(el)
                mockAvailable.forEach((item) => {
                        const option: HTMLElement = getByText(item.name)
                        expect(option).toBeInTheDocument()
                })
        })

        it('renders the selected options correctly', () => {
                const { getByText } = render(el)
                mockSelected.forEach((item) => {
                        const option: HTMLElement = getByText(item.name)
                        expect(option).toBeInTheDocument()
                })
        })

        it('should call handleChange everytime the available options is clicked', () => {
                const { getByLabelText } = render(el)
                const option: HTMLElement = getByLabelText(mockHeaderText)
                mockAvailable.forEach((item) => {
                        fireEvent.change(option, {
                                target: { value: item.name },
                        })
                        expect(calledAvailable).toContain(item)
                })
        })

        it('matches snapshot', () => {
                const tree = renderer.create().toJSON()
                expect(tree).toMatchSnapshot()
        })
})
