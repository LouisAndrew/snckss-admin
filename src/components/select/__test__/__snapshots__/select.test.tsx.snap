// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select reusable component matches snapshot 1`] = `null`;
